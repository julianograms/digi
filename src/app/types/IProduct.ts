export interface IProduct {
  name: string;
  detail: string;
  price: number;
  info?: string;
  image: string;
  hero?: string;
  offer?: string;
}
