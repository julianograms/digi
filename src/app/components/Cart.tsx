import Image from "next/image";

import { CartItem, useCartContext } from "../context/CartContext";
import CartListItem from "./CartListItem";
import { useMemo } from "react";

export default function Cart() {
  const { toggleCart, products } = useCartContext();
  const emptyList = useMemo(
    () =>
      !products.length ? (
        <div className="text-center mt-4 mb-4 text-lg text-gray-600">Empty</div>
      ) : null,
    [products.length]
  );
  const totalPrice = products.reduce(
    (acc: number, product: CartItem) =>
      acc + product.count * product.item.price,
    0
  );

  const cartList = products.map((product: CartItem) => (
    <CartListItem key={product.item._id} product={product} />
  ));
  return (
    <div
      className="fixed right-0 h-screen md:h-fit w-full md:w-[25rem] z-10  bg-white border border-black"
      data-testid="cart-modal"
    >
      <div className="h-16 flex items-center justify-between  border-b-2 border-black">
        <div className="text-black text-2xl pl-5">CART</div>
        <button onClick={toggleCart}>
          <Image src="/close.png" width={40} height={40} alt={"cart"} />
        </button>
      </div>
      <div className="flex flex-col w-full">
        {emptyList}
        {cartList}
      </div>
      <div className="flex text-2xl mt-4 mb-4">
        <div className="ml-auto pl-6 pr-2">Total:</div>
        <div>{`$${totalPrice.toFixed(2)}`}</div>
      </div>
    </div>
  );
}
