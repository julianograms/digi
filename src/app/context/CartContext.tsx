import { useContext, createContext, useState } from "react";
import { IProduct } from "../types/IProduct";
import Cart from "../components/Cart";

export type CartItem = { item: IProduct; count: number };

type CartContextType = {
  addToCart: (product: IProduct) => void;
  toggleCart: () => void;
  products: CartItem[];
};

export const CartContext = createContext({} as CartContextType);

export const useCartContext = () => {
  const context = useContext(CartContext);

  if (!context) throw new Error("No provider");

  return context;
};

type Props = {
  children: React.ReactNode;
};

export function CartProvider({ children }: Props) {
  const [products, setProducts] = useState<CartItem[]>([]);
  const [showCart, setShowCart] = useState(false);

  const addToCart = (productToAdd: IProduct) => {
    const foundIndex = products.findIndex(
      (product) => product.item._id === productToAdd._id
    );
    const isFound = foundIndex !== -1;
    if (isFound) {
      products[foundIndex].count++;
      setProducts([...products]);
    } else {
      setProducts([
        ...products,
        {
          item: productToAdd,
          count: 1,
        },
      ]);
    }
  };

  const toggleCart = () => {
    setShowCart(!showCart);
  };
  return (
    <CartContext.Provider value={{ products, addToCart, toggleCart }}>
      {showCart ? <Cart /> : null}
      {children}
    </CartContext.Provider>
  );
}
