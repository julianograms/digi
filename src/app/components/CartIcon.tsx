"use client";
import Image from "next/image";
import { CartItem, useCartContext } from "../context/CartContext";
import { useMemo } from "react";

export default function CartIcon() {
  const { products, toggleCart } = useCartContext();
  const count = products.reduce((acc: number, product: CartItem) => {
    return product.count + acc;
  }, 0);

  const countIconSize = useMemo(() => {
    if (count >= 10) {
      return "w-8";
    } else {
      return "w-[1.55rem]";
    }
  }, [count]);

  return (
    <button
      className="relative bg-white h-full flex items-center content-center"
      onClick={toggleCart}
    >
      {count ? (
        <div
          className={`absolute mb-8 ml-auto mr-auto ${countIconSize} left-0 right-0 bg-red-600 rounded-full text-center`}
        >
          <div data-testid="cart-counter">{count}</div>
        </div>
      ) : null}

      <div className={`${count ? "mt-4" : ""}`}>
        <Image src="/cart.png" width={40} height={40} alt={"cart"} />
      </div>
    </button>
  );
}
