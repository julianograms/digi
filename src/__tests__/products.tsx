import { fireEvent, render } from "@testing-library/react";
import { CartContext, CartProvider } from "@/app/context/CartContext";
import CartIcon from "@/app/components/CartIcon";
import ListItem from "@/app/components/ProductItem";
import Cart from "@/app/components/Cart";
import { IProduct } from "@/app/types/IProduct";

it("Show cart count from context", () => {
  const cartItems = [
    { count: 34, item: { name: "", _id: 0, image: "", price: 0, detail: "" } },
  ];

  const element = render(
    <CartContext.Provider
      value={{ addToCart: () => {}, products: cartItems, toggleCart: () => {} }}
    >
      <CartIcon />
    </CartContext.Provider>
  );

  expect(element.baseElement).toMatchSnapshot();
});

it("Should show the count number on the navbar icon", async () => {
  const element = render(
    <CartProvider>
      <ListItem
        item={{
          name: "",
          _id: 0,
          image: "http://placehold.it/300x300/999/CCC",
          price: 0,
          detail: "",
        }}
      />
      <CartIcon />
    </CartProvider>
  );

  const button = await element.findByText("Add to cart");
  fireEvent.click(button);
  const cartCounter = await element.findByTestId("cart-counter");

  expect(cartCounter.textContent).toBe("1");
});

it("Should show the cart products after click", async () => {
  const element = render(
    <CartProvider>
      <ListItem
        item={{
          name: "",
          _id: 0,
          image: "http://placehold.it/300x300/999/CCC",
          price: 0,
          detail: "",
        }}
      />
      <ListItem
        item={{
          name: "",
          _id: 1,
          image: "http://placehold.it/300x300/999/CCC",
          price: 0,
          detail: "",
        }}
      />
      <Cart />
    </CartProvider>
  );

  const [button1, button2] = await element.findAllByText("Add to cart");
  fireEvent.click(button1);
  fireEvent.click(button2);
  const cartCounter = await element.findByTestId("cart-modal");

  expect(cartCounter).toMatchSnapshot();
});

it("Should show offer", () => {
  const product: IProduct = {
    name: "",
    _id: 0,
    image: "http://placehold.it/300x300/999/CCC",
    price: 0,
    detail: "",
    offer: "Offer",
  };

  const element = render(<ListItem item={product} />);

  expect(element.baseElement).toMatchSnapshot();
});
