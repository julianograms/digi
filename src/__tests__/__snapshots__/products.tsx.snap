// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should show offer 1`] = `
<body>
  <div>
    <div
      class="flex flex-col  border-2 border-black w-104 h-[500px] mb-5"
    >
      <div
        class="w-[300px] h-[300px] flex items-center justify-center border-b-2 mb-4"
      >
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="300"
          src=""
          style="color: transparent;"
          width="300"
        />
      </div>
      <div
        class="w-64 pl-6 flex flex-col flex-1"
      >
        <div>
          <span
            class="text-lg font-bold"
          />
        </div>
        <div
          class="mb-3"
        />
        <div
          class="flex flex-row items-center content-center "
        >
          <img
            alt=""
            data-nimg="1"
            decoding="async"
            height="30"
            loading="lazy"
            src="/_next/image?url=%2Foffer.png&w=64&q=75"
            srcset="/_next/image?url=%2Foffer.png&w=32&q=75 1x, /_next/image?url=%2Foffer.png&w=64&q=75 2x"
            style="color: transparent;"
            width="30"
          />
          <div
            class="pl-3"
          >
            Offer
          </div>
        </div>
      </div>
      <div
        class="self-baseline w-full"
      >
        <div
          class="text-right text-2xl text-emerald-700"
        >
          $ 0.00
        </div>
        <button
          class=" text-center w-full justify-center p-4 bg-black text-white"
        >
          Add to cart
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Should show the cart products after click 1`] = `
<div
  class="fixed right-0 h-screen md:h-fit w-full md:w-[25rem] z-10  bg-white border border-black"
  data-testid="cart-modal"
>
  <div
    class="h-16 flex items-center justify-between  border-b-2 border-black"
  >
    <div
      class="text-black text-2xl pl-5"
    >
      CART
    </div>
    <button>
      <img
        alt="cart"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fclose.png&w=96&q=75"
        srcset="/_next/image?url=%2Fclose.png&w=48&q=75 1x, /_next/image?url=%2Fclose.png&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
    </button>
  </div>
  <div
    class="flex flex-col w-full"
  >
    <div
      class="border-b w-full cart-list-grid"
    >
      <div
        class="w-[40px] h-[40px] flex items-center justify-center"
      >
        <img
          alt="cart"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=http%3A%2F%2Fplacehold.it%2F300x300%2F999%2FCCC&w=96&q=75"
          srcset="/_next/image?url=http%3A%2F%2Fplacehold.it%2F300x300%2F999%2FCCC&w=48&q=75 1x, /_next/image?url=http%3A%2F%2Fplacehold.it%2F300x300%2F999%2FCCC&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </div>
      <div
        class="flex items-center content-center pl-2"
      >
        <div
          class="text-center"
        />
      </div>
      <div
        class="p-2 bg-gray-300 text-center"
      >
        1
      </div>
      <div
        class="flex items-center content-center"
      >
        <div
          class="ml-auto"
        >
          $ 0.00
        </div>
      </div>
    </div>
    <div
      class="border-b w-full cart-list-grid"
    >
      <div
        class="w-[40px] h-[40px] flex items-center justify-center"
      >
        <img
          alt="cart"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=http%3A%2F%2Fplacehold.it%2F300x300%2F999%2FCCC&w=96&q=75"
          srcset="/_next/image?url=http%3A%2F%2Fplacehold.it%2F300x300%2F999%2FCCC&w=48&q=75 1x, /_next/image?url=http%3A%2F%2Fplacehold.it%2F300x300%2F999%2FCCC&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </div>
      <div
        class="flex items-center content-center pl-2"
      >
        <div
          class="text-center"
        />
      </div>
      <div
        class="p-2 bg-gray-300 text-center"
      >
        1
      </div>
      <div
        class="flex items-center content-center"
      >
        <div
          class="ml-auto"
        >
          $ 0.00
        </div>
      </div>
    </div>
  </div>
  <div
    class="flex text-2xl mt-4 mb-4"
  >
    <div
      class="ml-auto pl-6 pr-2"
    >
      Total:
    </div>
    <div>
      $0.00
    </div>
  </div>
</div>
`;

exports[`Show cart count from context 1`] = `
<body>
  <div>
    <button
      class="relative bg-white h-full flex items-center content-center"
    >
      <div
        class="absolute mb-8 ml-auto mr-auto w-8 left-0 right-0 bg-red-600 rounded-full text-center"
      >
        <div
          data-testid="cart-counter"
        >
          34
        </div>
      </div>
      <div
        class="mt-4"
      >
        <img
          alt="cart"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fcart.png&w=96&q=75"
          srcset="/_next/image?url=%2Fcart.png&w=48&q=75 1x, /_next/image?url=%2Fcart.png&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </div>
    </button>
  </div>
</body>
`;
